torch/csrc/distributed/autograd/context/DistAutogradContainer.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <mutex>
+#include <unordered_map>
+
+#include <torch/csrc/distributed/autograd/context/DistAutogradContext.h>
+
+namespace torch {
+namespace distributed {
+namespace autograd {
+
+/*
+  Singleton class per worker which is responsible for storing the distributed
+  autograd context for each autograd pass and also cleans up data for an
+  autograd pass once its done.
+
+  Each autograd pass is assinged a unique autograd_context_id and all data for
+  that pass (DistAutogradContext) is stored in this container indexed by the
+  autograd_context_id. The autograd_context_id itself is a 64 bit globally
+  unique id. The first 16 bits is the worker_id and the next 48 bits is an
+  auto-incrementing id for each worker.
+*/
+class DistAutogradContainer {
+ public:
+  static DistAutogradContainer& init(int16_t worker_id);
+  static DistAutogradContainer& getInstance();
+  const DistAutogradContext& newContext();
+  void releaseContext(int64_t context_id);
+  const DistAutogradContext& retrieveContext(int64_t context_id);
+
+ private:
+  DistAutogradContainer();
+  ~DistAutogradContainer() = default;
+
+  DistAutogradContainer(const DistAutogradContainer&) = delete;
+  DistAutogradContainer& operator=(const DistAutogradContainer&) = delete;
+  DistAutogradContainer(DistAutogradContainer&&) = delete;
+  DistAutogradContainer& operator=(DistAutogradContainer&&) = delete;
+
+  // Auto incrementing context id used to identify unique autograd passes.
+  // Initialized with the first 16 bits being the worker_id.
+  int64_t current_context_id_;
+
+  // Unique id to identify a worker in the distributed setting.
+  int16_t worker_id_;
```

Comment:
If using ProcessGroup, will this be the rank? I believe rank is `int` in `ProcessGroup`, any reason for truncate it to 16 bits?